### PR TITLE
implement read_multi_entry to work alongside read_entry/write_entry

### PR DESCRIPTION
**Pull request by grosser**
- reduce duplication of code between mem_cache_store and base
- make it trivial to add new stores that define their own read_multi

caches that inherit cache can use read_multi_entry but it won't break them if they currently define read_multi

@rafaelfranca

### CODE REVIEW

**Comment by grosser** [comment]
@rafaelfranca @kaspth this is good to go ?
no fallout afaik ...

File: activesupport/lib/active_support/cache.rb [hunk 1]
```diff
@@ -333,19 +333,16 @@ def read_multi(*names)
         options = merged_options(options)
 
         instrument_multi(:read, names, options) do |payload|
-          results = {}
-          names.each do |name|
-            key = normalize_key(name, options)
-            entry = read_entry(key, options)
+          keys_to_names = Hash[names.map { |name| [normalize_key(name, options), name] }]
+          read_multi_entry(keys_to_names.keys, options).each_with_object({}) do |(key, entry), results|
```
Comment (kaspth) [inline activesupport/lib/active_support/cache.rb]: `each_with_object` is called just after another `each_with_object` in both `read_multi_entry` calls, which is to say, we're looping over the collection several times.

What's the performance implications of this? I'd love to see a benchmark so we know we don't have a regression here :smile:

**Comment by schneems** [comment]
I'm fine with the concept. I think it is useful. Do we have any documentation or guides for someone who is maintaining a Rails.cache backend? I think we should specify how to add support to the method overwrite similar to the memcache backend.

Also it looks like `dalli` already has a `read_multi`. Can we use preserve that as a method name or is there concern for arity problems? https://github.com/mperham/dalli/blob/172176c547ee5716e6236efbe9885a2b0b1b3ab2/lib/active_support/cache/dalli_store.rb#L168-L192 cc/ @mperham

**Comment by mperham** [comment]
@schneems I'm no longer maintaining Dalli. @petergoldstein is the new maintainer.

**Comment by schneems** [comment]
Ahh, thanks for the heads up. Also thanks for taking up maintainance @petergoldstein.

**Comment by petergoldstein** [comment]
@schneems Happy to do it.

Regarding this change, I agree that it looks fine conceptually.

Regarding the method name, I think `read_multi_entry` is a better choice to avoid arity conflicts with the existing `read_multi` method on the ActiveSupport::Cache::DalliStore.  I suspect that @grosser would be willing to commit a PR with the new method on DalliStore.  That'll make upgrades simple.

**Comment by grosser** [comment]
Yeah, I'd try to get that into dalli / libmemcached_store first.
I think it should be named read_multi_entry to make it symmetric to all the
other methods and not add more 'key vs name' confusion.

**Comment by kamipo** [comment]
Closing in favor of #30268.

**Pull request closed without merging**